webhook/resources/backuptarget/validator.go

```diff
@@ -0,0 +1,221 @@
+	oldBackupTarget := oldObj.(*longhorn.BackupTarget)
+	newBackupTarget := newObj.(*longhorn.BackupTarget)
+
+	urlChanged := oldBackupTarget.Spec.BackupTargetURL != newBackupTarget.Spec.BackupTargetURL
+	secretChanged := oldBackupTarget.Spec.CredentialSecret != newBackupTarget.Spec.CredentialSecret
+
+	if urlChanged {
+		if err := b.ds.ValidateBackupTargetURL(newBackupTarget.Spec.BackupTargetURL); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if secretChanged {
+		if err := b.validateCredentialSecret(newBackupTarget.Spec.CredentialSecret); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if urlChanged || secretChanged {
+		if err := b.validateDRVolume(newBackupTarget); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	return nil
+}
+
+func (b *backupTargetValidator) validateCredentialSecret(secretName string) error {
+	namespace, err := b.ds.GetLonghornNamespace()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get Longhorn namespace")
+	}
+	secret, err := b.ds.GetSecretRO(namespace.Name, secretName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to get the secret before modifying backup target credential secret")
+		}
+		return nil
+	}
+	checkKeyList := []string{
+		types.AWSIAMRoleAnnotation,
+		types.AWSIAMRoleArn,
+		types.AWSAccessKey,
+		types.AWSSecretKey,
+		types.AWSEndPoint,
+		types.AWSCert,
+		types.CIFSUsername,
+		types.CIFSPassword,
+		types.AZBlobAccountName,
+		types.AZBlobAccountKey,
+		types.AZBlobEndpoint,
+		types.AZBlobCert,
+		types.HTTPSProxy,
+		types.HTTPProxy,
+		types.NOProxy,
+		types.VirtualHostedStyle,
+	}
+	for _, checkKey := range checkKeyList {
+		if value, ok := secret.Data[checkKey]; ok {
+			if strings.TrimSpace(string(value)) != string(value) {
+				return fmt.Errorf("there is space or new line in %s", checkKey)
+			}
+		}
+	}
+	return nil
+}
+
+func (b *backupTargetValidator) validateDRVolume(backupTarget *longhorn.BackupTarget) error {
+	vs, err := b.ds.ListDRVolumesRO()
+	if err != nil {
+		return errors.Wrapf(err, "failed to list standby volume when modifying BackupTarget")
+	}
+
+	backupTargetURL, err := getBackupTargetURLWithoutOpts(backupTarget.Spec.BackupTargetURL)
+	if err != nil {
+		return err
+	}
+
+	standbyVolumeNames := sets.NewString()
+	for k, v := range vs {
+		backupTargetURLfromBackupURL, err := getBackupTargetURLWithoutOpts(v.Spec.FromBackup)
+		if err != nil {
+			return err
+		}
+		fromBackupTargetName, ok := v.Labels[types.LonghornLabelBackupTarget]
+		if (ok && fromBackupTargetName == backupTarget.Name) || backupTargetURLfromBackupURL == backupTargetURL {
+			standbyVolumeNames.Insert(k)
+		}
+	}
+	if standbyVolumeNames.Len() != 0 {
+		return fmt.Errorf("cannot modify BackupTarget since there are existing standby volumes: %v", standbyVolumeNames)
+	}
```

Comment:
nit: we don't need to check all volumes. If any DR is volume, return an error directly.

Reply:
Modified. Thanks.